### DISCUSSION TITLE
Maximum length of value in an AutoConnectInput?

### DISCUSSION THREAD

**Discussion by ac1176** [Hieromon/AutoConnect#405] [Q&A]
Hello again Hieromon,

Do you know if there is an upper limit to the length of text that a user can enter into an AutoConnectInput field when submitting an online form created in AutoConnect? I have an AutoConnectInput configured on a custom page similar to as shown on the [Input Data Validation](https://hieromon.github.io/AutoConnect/achandling.html#input-data-validation) examples with a validation pattern set as any character up to a maximum of 40 in length (regex `^.{1,40}$`), but I'm finding that any text entered in the field longer than 24 characters causes the ESP32 to crash with a `Guru Meditation Error: Core  1 panic'ed (Unhandled debug exception) Debug exception reason: Stack canary watchpoint triggered (loopTask)` error.

If the user enters the text `This is a small group of words` (24 characters) and presses the 'Save' button the program passes the validation and executes properly. If the user enters `This is a small group of wordss` (25 characters) and presses the 'Save button the ESP32 crashes, despite the regex pattern validation still being true.

Abbreviated code below:

```
ACInput(inputName, "", "Name", "^.{1,40}$", "Name goes here");
ACSubmit(savecfg, "Save", "/save_settings");
AutoConnectAux test_page("/test_page", "Test Page", true, {
   inputName,
   savecfg
});

...

ACText(caption, "");
AutoConnectAux save_settings("/save_settings", "Save Page", false, {
  caption
});

...

String saveParams(AutoConnectAux& savePage, PageArgument& args) {
   AutoConnectAux& inputCfg = *Portal.aux("/test_page");
   AutoConnectInput& enteredName = inputCfg["inputName"].as<AutoConnectInput>();
   AutoConnectText& result = savePage["caption"].as<AutoConnectText>();
   if (enteredName.isValid()){

    //do something with submitted value
   result.value = "<h4>Name accepted</h4>";

  }
  else{
    
    //reject the submitted value
    result.value = "<h4>Name rejected, too long</h4>";

  }
  return String("");
}

...

void setup(){
  Portal.join({test_page, save_settings});
  Portal.on("/save_settings", saveParams);

...etc
```

Thanks!

**Comment by ac1176** [comment]
I should really spend some more time researching before asking, shouldn't I? 😉 

Other sources online suggest that the `Stack canary watchpoint triggered (loopTask)` error can be triggered if there isn't enough stack memory assigned on program compilation. If I increase the stack memory allocation in my Arduino sdkconfig.h file (#define ARDUINO_LOOP_STACK_SIZE) from the default 8192 to 16384 the error/crash goes away and the pattern validation can correctly detect the 40 character limit.

**Reply by Hieromon** [reply]
@ac1176 Hello,
You have finally come across this problem. I've spent a lot of time suppressing this problem, but haven't found a complete solution.

It is the ESP-IDF **std::regex** library that creates the problem.
ESP-IDF is based on the xtensa-gcc and is c++17 compliant since 4.0. As you pointed out, the stack size required for std::regex is 8KB or more. Therefore, a small loop thread stack size will raise an exception.
And the workaround I have come up with is stupid, get it triggered by enabling the `AUTOCONNECT_UPLOAD_ASFIRMWARE_USE_REGEXP` macro.

https://github.com/Hieromon/AutoConnect/blob/e9243476de64f7ab40bce84bb48bf2364e975ce2/src/AutoConnectOTA.cpp#L179-L194

The above logic is an implementation of [this topic]( https://github.com/Hieromon/AutoConnect/issues/236).  I internally created a dedicated thread to give std::regex the required amount of stack. Of course, this way is not appropriate. If the user sketch creates its own separate thread, it will probably raise an exception. So I sealed this.
At this point, I was aware that the same thing would happen with AutoConnectInput validation, but I was hoping that the ESP-IDF side would do something about it. So, I left the problem for the time being. You have stepped on a mine.

Anyway, I'll continue to verify whether the problem can be reproduced even with esp32 core 2.0.0. I would be grateful if you could let me know if you find any countermeasures.
Thank you.

**Reply by ac1176** [reply]
Sorry to hear that. At least I've managed to unwittingly become the canary in this situation 😁 

For what its worth, I noticed that ESP V2.0.0 includes [a new feature](https://github.com/espressif/arduino-esp32/pull/5173) that makes it possible to override the stack size on compilation with an arbitrary value. But I can't find any documentation describing how to actually use it within the Arduino IDE. Even more annoying, if I upgrade to ESPV2.0.0 the Arduino IDE will no longer compile my program for the NODEMCU-32S board I'm using, so unfortunately I'm stuck using V1.0.6 and my manual hacking of the stack size within sdkconfig.h.

To be fair, my temporary fix does seem to work fine and the code appears to function OK. Doubling the stack size is probably unnecessary and extreme; I haven't done any further experiments to find out exactly how much stack I really need, so I guess there's every possibility I only need a few extra words above the default 8k limit. But I suppose there's also nothing stopping the user entering hundreds of characters in an AutoConnectInput field, exceeding a higher stack limit and crashing the ESP32 anyway.
